File: 2022/October/pg_2d_coin_flip.js
const solution = (beginning, target) => {
  let cntCase = 0;

  let standRow = "";

  for (let i = 0; i < beginning.length; i++) {
    let crntRow = "";
    for (let j = 0; j < beginning[0].length; j++) {
      if (beginning[i][j] !== target[i][j]) {
        crntRow += "1";

        // 세로 축 기준으로 합산
        if (i === 0) {
          cntCase += 1;
        } else {
          // 가로축 기준으로 합산
          if (standRow[0] === "0" && j === 0) {
            cntCase += 1;
          }
        }
      } else {
        crntRow += "0";
        if (i > 0 && standRow[0] === "1" && j === 0) {
          cntCase += 1;
        }
      }
    }

    // 첫째 줄의 경우 기준 줄로 만들어준다.
    if (i === 0) {
      standRow = crntRow;
    } else {
      // 같지 않을 때
      if (standRow !== crntRow) {
        // if (standRow[0] !== crntRow[0]) {
        // 둘을 합쳐 모두 1(바꿔야 되는 자리)로 만들 수 있는 경우 -1을 반환한다.
        if (!(+standRow + +crntRow === Number("1".repeat(standRow.length)))) {
          return -1;
        }
      }
    }
  }

  return Math.min(cntCase, target.length + target[0].length - cntCase);
};

console.log(
  solution(
    [
      [0, 0],
      [0, 0],
    ],
    [
      [1, 1],
      [1, 0],
    ]
  )
);

File: 2022/October/pg_2d_coin_flip.py
def solution(beginning, target):
    answer = 0
    stand_row = ""
    
    h = len(beginning)
    w = len(beginning[0])
    
    for i in range(h):
        crnt_row = ''
        for j in range(w):
            if (beginning[i][j] != target[i][j]):
                crnt_row += '1'
                
                if i == 0:
                    answer += 1
                else:
                    if stand_row[0] == '0' and j == 0:
                        answer += 1
            else:
                crnt_row += '0'
                print(stand_row)
                if i > 0 and stand_row[0] == '1' and j == 0:
                    answer += 1
    
        if i == 0:
            stand_row = crnt_row
        else:
            if stand_row != crnt_row:
                if not int(stand_row) + int(crnt_row) == int('1' * w):
                    return -1
                
    return min(answer, h + w - answer)

beginning = [[0, 1, 0, 0, 0], [1, 0, 1, 0, 1], [0, 1, 1, 1, 0], [1, 0, 1, 1, 0], [0, 1, 0, 1, 0]]
target = [[0, 0, 0, 1, 1], [0, 0, 0, 0, 1], [0, 0, 1, 0, 1], [0, 0, 0, 1, 0], [0, 0, 0, 0, 1]]
result = 5

print(solution(beginning, target))